src/plasma_manager.c

```diff
@@ -164,101 +383,271 @@ void read_object_chunk(event_loop *loop,
+  client_connection *client_conn = object_conn->client_conn;
+  LOG_DEBUG("Timer went off, %d tries left", object_conn->num_retries);
+  if (object_conn->num_retries > 0) {
+    request_transfer_from(client_conn, object_conn->object_id);
+    object_conn->num_retries--;
+    return MANAGER_TIMEOUT;
+  }
+  plasma_reply reply = {.object_id = object_conn->object_id, .has_object = 0};
+  send_client_reply(client_conn, &reply);
+  remove_object_connection(client_conn, object_conn);
+  return AE_NOMORE;
+}
+
+/**
+ * Given an object ID and the managers it can be found on, start requesting a
+ * transfer from the managers.
+ *
+ * @param object_id The object ID we want to request a transfer of.
+ * @param manager_count The number of managers the object can be found on.
+ * @param manager_vector A vector of the IP addresses of the managers that the
+ *        object can be found on.
+ * @param context The context for the connection to this client.
+ *
+ * Initializes a new context for this client and object. Managers are tried in
+ * order until we receive the data or we timeout and run out of retries.
+ */
+void request_transfer(object_id object_id,
+                      int manager_count,
+                      const char *manager_vector[],
+                      void *context) {
+  client_connection *client_conn = (client_connection *) context;
+  client_object_connection *object_conn =
+      get_object_connection(client_conn, object_id);
+  CHECK(object_conn);
+  LOG_DEBUG("Object is on %d managers", manager_count);
+  if (manager_count == 0) {
+    /* TODO(swang): Instead of immediately counting this as a failure, maybe
+     * register a Redis callback for changes to this object table entry. */
+    free(manager_vector);
+    plasma_reply reply = {.object_id = object_conn->object_id, .has_object = 0};
+    send_client_reply(client_conn, &reply);
+    remove_object_connection(client_conn, object_conn);
+    return;
+  }
+  /* Pick a different manager to request a transfer from on every attempt. */
+  object_conn->manager_count = manager_count;
+  object_conn->manager_vector = malloc(manager_count * sizeof(char *));
+  memset(object_conn->manager_vector, 0, manager_count * sizeof(char *));
+  for (int i = 0; i < manager_count; i++) {
+    int len = strlen(manager_vector[i]);
+    object_conn->manager_vector[i] = malloc(len + 1);
+    strncpy(object_conn->manager_vector[i], manager_vector[i], len);
+    object_conn->manager_vector[i][len] = '\0';
+  }
+  free(manager_vector);
+  /* Wait for the object data for the default number of retries, which timeout
+   * after a default interval. */
+  object_conn->num_retries = NUM_RETRIES;
+  object_conn->timer =
+      event_loop_add_timer(client_conn->manager_state->loop, MANAGER_TIMEOUT,
+                           manager_timeout_handler, object_conn);
+  request_transfer_from(client_conn, object_id);
+}
+
+void process_fetch_request(client_connection *client_conn,
```

Comment:
Consider documenting the logic of when `send_client_reply` will be called? If I understand correctly, it is something like.

"This method will reply to the client that called fetch if the object is available locally or if the manager is not connected to the database. Otherwise, it will call object_table_lookup, which queries the database for the locations of the objects. When the object table responds, the callback request_transfer is called, which will initiate a transfer from another plasma manager. If the transfer succeeds, then the reply to the client will be sent in process_data_chunk when the transfer finishes. Otherwise, the transfer will timeout and we will reply to the client in the callback manager_timeout_handler."